x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termsenum/action/TransportTermsEnumAction.java

```diff
@@ -431,21 +432,29 @@ private boolean canAccess(
 
                 // Current user has potentially many roles and therefore potentially many queries
                 // defining sets of docs accessible
-                Set<BytesReference> queries = indexAccessControl.getDocumentPermissions().getQueries();
-                for (BytesReference querySource : queries) {
-                    QueryBuilder queryBuilder = DLSRoleQueryValidator.evaluateAndVerifyRoleQuery(
-                        querySource,
-                        scriptService,
-                        queryShardContext.getParserConfig().registry(),
-                        securityContext.getUser()
-                    );
-                    QueryBuilder rewrittenQueryBuilder = Rewriteable.rewrite(queryBuilder, queryShardContext);
-                    if (rewrittenQueryBuilder instanceof MatchAllQueryBuilder) {
-                        // One of the roles assigned has "all" permissions - allow unfettered access to termsDict
-                        return true;
+                final List<Set<BytesReference>> listOfRawQueries = indexAccessControl.getDocumentPermissions().getListOfQueries();
+
+                return listOfRawQueries.stream().allMatch(queries -> {
+                    for (BytesReference querySource : queries) {
+                        QueryBuilder queryBuilder = DLSRoleQueryValidator.evaluateAndVerifyRoleQuery(
+                            querySource,
+                            scriptService,
+                            queryShardContext.getParserConfig().registry(),
+                            securityContext.getUser()
+                        );
+                        QueryBuilder rewrittenQueryBuilder;
+                        try {
+                            rewrittenQueryBuilder = Rewriteable.rewrite(queryBuilder, queryShardContext);
+                        } catch (IOException e) {
+                            throw new UncheckedIOException(e);
+                        }
+                        if (rewrittenQueryBuilder instanceof MatchAllQueryBuilder) {
+                            // One of the roles assigned has "all" permissions - allow unfettered access to termsDict
+                            return true;
+                        }
                     }
-                }
-                return false;
+                    return false;
+                });
```

Comment:
The change here is also a bug fix. The existing code only checks `getQueries()` but not `getLimitedByQueries`. So it can make wrong decisions if the authentication subject is an API key. It is possible to maitain the existing behaviour so that the PR is more of a pure refactoring. But with the changes to DocumentPermissions, maintaining the existing behaviour is rather odd because all queries are now return by a single method.

Reply:
I raised a separate PR for the bug fix https://github.com/elastic/elasticsearch/pull/91170 to keep this PR a pure refactoring.